### PR TITLE
fix lm head overriden issue, move it from checkpoint in-loop loading …

### PR DESCRIPTION
**Pull request by sywangyi**
…to out loop

### CODE REVIEW

**Comment by delock** [comment]
What problem caused by inner loop loading? @sywangyi

**Comment by sywangyi** [comment]
r_module.lm_head.weight will be wrongly pointed to word_embeddings buffer since it's in the last sharded checkpoint. acturally in llama70b, there's lm_head weight in the checkpoint. so lm head and work embedding should point to different buffers.

**Comment by yao-matrix** [comment]
This issue will lead to llama2-70B model generate wrong texts in both CPU and Habana Gaudi.

**Comment by delock** [comment]
GPT-J would also have problem when checkpoints are locally sharded, and could be fixed by this PR.

**Comment by delock** [comment]
@sywangyi Your PR can fix a correctness bug in Llama2-7b when loaded with from_config, can you rebase to master and resolve conflcit?

Hi @RezaYazdaniAminabadi, could this PR be merged into master?  There is a bug in master branch with Llama2 from_config loading and can be fixed by this PR.  We will open an issue soon.

File: deepspeed/module_inject/replace_module.py [hunk 1]
```diff
@@ -386,6 +393,13 @@ def replace_fn(child, _policy, layer_id=0, prefix="", state_dict=None):
                                                container=container_g)
                     sds = [None for _ in sds]
                     gc.collect()
+        embedding_weight = None
```
Comment (tjruwase) [inline deepspeed/module_inject/replace_module.py:396]: Could this be refactored into a function? It looks like duplicate of other change.

Comment (sywangyi) [reply deepspeed/module_inject/replace_module.py]: done

**Review by tjruwase — APPROVED** [review]

**Pull request merged**